### PR TITLE
fix glob matching of alternatives

### PR DESCRIPTION
**Pull request by sokra**
### Description

The unicode refactoring introduced some bugs in matching of alternatives

Closes PACK-2172

### CODE REVIEW

File: crates/turbo-tasks-fs/src/glob.rs [hunk 1]
```diff
@@ -347,28 +340,25 @@ impl<'a> Iterator for GlobPartMatchesIterator<'a> {
...
-                let end = match self.cursor.next_boundary(self.path, 0) {
-                    Ok(Some(end)) => end,
-                    _ => return None,
+            GlobPart::FileChar(chars) => {
+                let start = self.cursor.cur_cursor();
+                let Ok(Some(end)) = self.cursor.next_boundary(self.path, 0) else {
+                    return None;
                 };
-
-                let c = &chars[self.cursor.cur_cursor()..end]
```
Comment (sokra) [inline crates/turbo-tasks-fs/src/glob.rs:356]: This is unused, but it doesn't really make sense to use the path cursor to index `chars`

File: crates/turbo-tasks-fs/src/glob.rs [hunk 2]
```diff
@@ -379,15 +369,10 @@ impl<'a> Iterator for GlobPartMatchesIterator<'a> {
...
-                        let end = self.cursor.next_boundary(self.path, 0);
+                        self.index += 1;
                         self.glob_iterator = None;
-                        match end {
-                            Ok(Some(_)) => {}
-                            Ok(None) => return None,
-                            Err(..) => return None,
-                        }
                     }
-                } else if let Some(alternative) = alternatives.get(self.cursor.cur_cursor()) {
```
Comment (sokra) [inline crates/turbo-tasks-fs/src/glob.rs:390]: path cursor can't be used to index alternatives list

File: crates/turbo-tasks-fs/src/glob.rs [hunk 3]
```diff
@@ -267,12 +269,9 @@ impl<'a> Iterator for GlobPartMatchesIterator<'a> {
...
                 if self.cursor.cur_cursor() == 0 {
-                    let end = self.cursor.next_boundary(self.path, 0);
-                    match end {
-                        Ok(Some(_)) => {}
-                        Ok(None) => return None,
-                        Err(..) => return None,
-                    }
+                    let Ok(Some(_)) = self.cursor.next_boundary(self.path, 0) else {
+                        return None;
+                    };
```
Comment (sokra) [inline crates/turbo-tasks-fs/src/glob.rs:274]: This is just shorter and cleaner

Comment (kwonoj) [reply crates/turbo-tasks-fs/src/glob.rs:274]: yes I'm always forgetting there's let..else syntax

File: crates/turbo-tasks-fs/src/glob.rs [hunk 4]
```diff
@@ -457,19 +442,43 @@ mod tests {
...
+         sharp,shiki,sqlite3,tailwindcss,ts-node,typescript,vscode-oniguruma,webpack,websocket,@\
+         aws-sdk/client-dynamodb,@aws-sdk/lib-dynamodb}/**",
+        "node_modules/.pnpm/@aws-sdk+lib-dynamodb@3.445.0_@aws-sdk+client-dynamodb@3.445.0/\
+         node_modules/@aws-sdk/lib-dynamodb/dist-es/index.js"
+    )]
     #[case::alternatives_nested1("{a,b/c,d/e/{f,g/h}}", "a")]
     #[case::alternatives_nested2("{a,b/c,d/e/{f,g/h}}", "b/c")]
     #[case::alternatives_nested3("{a,b/c,d/e/{f,g/h}}", "d/e/f")]
     #[case::alternatives_nested4("{a,b/c,d/e/{f,g/h}}", "d/e/g/h")]
+    // #[case::alternatives_chars("[abc]", "b")]
```
Comment (kwonoj) [inline crates/turbo-tasks-fs/src/glob.rs:481]: is this inteded to be commented?

Comment (sokra) [reply crates/turbo-tasks-fs/src/glob.rs:481]: This isn't supported in the glob parsing yet. We might want to uncomment that test case once we implement that.

**Review by kdy1 — APPROVED** [review]

**Pull request merged**